File: generate_qr.py
import qrcode
from PIL import Image
import os

# URL to encode
url = "https://www.822athletics.com"

# Create QR code instance
qr = qrcode.QRCode(
    version=1,
    error_correction=qrcode.constants.ERROR_CORRECT_H,  # High error correction for logo overlay
    box_size=10,
    border=4,
)

# Add data
qr.add_data(url)
qr.make(fit=True)

# Create QR code image
qr_img = qr.make_image(fill_color="black", back_color="white").convert('RGB')

# Open logo image
logo_path = r"C:\Users\deltabro\Downloads\Photo Apr 03 2019, 10 56 53 AM.png"
logo = Image.open(logo_path)

# Calculate logo size (roughly 1/5 of QR code size)
qr_width, qr_height = qr_img.size
logo_size = qr_width // 5

# Resize logo
logo = logo.resize((logo_size, logo_size), Image.Resampling.LANCZOS)

# Calculate position to center logo
logo_pos = ((qr_width - logo_size) // 2, (qr_height - logo_size) // 2)

# Paste logo onto QR code
qr_img.paste(logo, logo_pos, logo if logo.mode == 'RGBA' else None)

# Save the final QR code
output_path = "822athletics_qr_code.png"
qr_img.save(output_path)

print(f"QR code generated successfully: {output_path}")

File: generate_qr.js
const QRCode = require('qrcode');
const sharp = require('sharp');
const fs = require('fs');

async function generateQRWithLogo() {
  try {
    const url = 'https://www.822athletics.com';
    const logoPath = 'C:\\Users\\deltabro\\Downloads\\Photo Apr 03 2019, 10 56 53 AM.png';
    const outputPath = '822athletics_qr_code.png';

    // Generate QR code as buffer with high error correction
    const qrBuffer = await QRCode.toBuffer(url, {
      errorCorrectionLevel: 'H',
      type: 'png',
      width: 600,
      margin: 2,
      color: {
        dark: '#000000',
        light: '#FFFFFF'
      }
    });

    // Get QR code dimensions
    const qrImage = sharp(qrBuffer);
    const qrMetadata = await qrImage.metadata();

    // Calculate logo size (about 1/5 of QR code)
    const logoSize = Math.floor(qrMetadata.width / 5);

    // Create white square with padding - make it align properly to avoid black pixels showing
    const whiteSquareSize = Math.floor(qrMetadata.width / 3.2); // Slightly larger for clean edges
    const whiteSquare = await sharp({
      create: {
        width: whiteSquareSize,
        height: whiteSquareSize,
        channels: 3,
        background: { r: 255, g: 255, b: 255 }
      }
    }).png().toBuffer();

    // Resize logo with white background
    const resizedLogo = await sharp(logoPath)
      .resize(logoSize, logoSize, {
        fit: 'contain',
        background: { r: 255, g: 255, b: 255, alpha: 1 }
      })
      .toBuffer();

    // Calculate position to center the white square
    const whiteSquarePosition = {
      left: Math.floor((qrMetadata.width - whiteSquareSize) / 2),
      top: Math.floor((qrMetadata.height - whiteSquareSize) / 2)
    };

    // Calculate position to center the logo
    const logoPosition = {
      left: Math.floor((qrMetadata.width - logoSize) / 2),
      top: Math.floor((qrMetadata.height - logoSize) / 2)
    };

    // Composite white square first, then logo on top
    const finalImage = await sharp(qrBuffer)
      .composite([
        {
          input: whiteSquare,
          left: whiteSquarePosition.left,
          top: whiteSquarePosition.top
        },
        {
          input: resizedLogo,
          left: logoPosition.left,
          top: logoPosition.top
        }
      ])
      .toFile(outputPath);

    console.log(`QR code generated successfully: ${outputPath}`);
    console.log(`Size: ${finalImage.width}x${finalImage.height}`);
  } catch (error) {
    console.error('Error generating QR code:', error);
  }
}

generateQRWithLogo();
